File: tests/typesafety/python/exercise_money_misuse.py
# expect: fails
# title: Usd plus Eur, rejected by the checker
# misuse-of: exercise_money_solution
from dataclasses import dataclass, replace


@dataclass(frozen=True)
class Usd:
    cents: int


@dataclass(frozen=True)
class Eur:
    cents: int


def add[M: (Usd, Eur)](a: M, b: M) -> M:
    return replace(a, cents=a.cents + b.cents)


def basket_total() -> Usd:
    return add(Usd(1999), Eur(500))

File: tests/typesafety/typescript/exercise_money_misuse.ts
// expect: fails

type Currency = "USD" | "EUR";

type Money<C extends Currency> = {
  readonly currency: C;
  readonly cents: number;
};

function money<C extends Currency>(currency: C, cents: number): Money<C> {
  return { currency, cents };
}

function add<C extends Currency>(a: Money<C>, b: Money<NoInfer<C>>): Money<C> {
  return { currency: a.currency, cents: a.cents + b.cents };
}

export function basketTotal(): Money<"USD"> {
  return add(money("USD", 1999), money("EUR", 500));
}
